### PR TITLE
Fix usb

### PR DESCRIPTION
**Pull request by av500**
(NOTE: this is a copy of the pull request sent to libusbhost)

the existing code had 4 issues which are now fixed:

1) after a CTRL REQUEST DATA in (e.g. when read the device descriptor) there was NO status packet sent. this basically means that the only working devices were the ones that ignore USB protocol errors and continue.

2) some USB devices (e.g. AKAI LPK25) have a max packet size of only 8 bytes, so even reading the 18 byte device descriptor fails. now only 8 bytes are read to get max packet size and then the full descriptor is read.

3) some USB devices (e.g. Arturia Beatstep Pro) have a TOGGLE error at their first DATA IN transaction, the state machine did not handle that and simply stopped, now a TOGGLE ERROR is treated like any other error and data can flow again

4) when iterating over the interfaces the drivers' analyze function was always passed the data buffer of the FIRST interface, this is now fixed and devices like AKAI MIDIMIX work

### CODE REVIEW

**Comment by SloBloLabs** [comment]
Vladimir was so kind to send me his patch and I can confirm Akai MPK mini finally working with this. #127 would be solved for me having the PR in stock firmware.

File: src/platform/stm32/libs/libusbhost/src/usbh_core.c [hunk 1]
```diff
@@ -134,18 +134,19 @@ static void device_register(void *descriptors, uint16_t descriptors_len, usbh_de
...
-			LOG_PRINTF("INTERFACE_DESCRIPTOR\n");
 			struct usb_interface_descriptor *iface = (void*)&buf[i];
 			device_info.ifaceClass = iface->bInterfaceClass;
 			device_info.ifaceSubClass = iface->bInterfaceSubClass;
 			device_info.ifaceProtocol = iface->bInterfaceProtocol;
+			LOG_PRINTF("INTERFACE_DESCRIPTOR  Cls %02X  Sub %02X  Prot %02X \n", iface->bInterfaceClass, iface->bInterfaceSubClass, iface->bInterfaceProtocol);
 			if (find_driver(dev, &device_info)) {
+				uint8_t *ibuf = (uint8_t *)iface;
 				int k = 0;
 				while (k < descriptors_len) {
```
Comment (westlicht) [inline src/platform/stm32/libs/libusbhost/src/usbh_core.c:145]: While before we always used data of the first interface, now we don't check for end of buffer correctly anymore, right? Should probably be `while (k < descriptors_len - i)` now.

File: src/platform/stm32/libs/libusbhost/src/usbh_core.c [hunk 2]
```diff
@@ -134,18 +134,19 @@ static void device_register(void *descriptors, uint16_t descriptors_len, usbh_de
...
-					desc_len = buf[k];
+					desc_len = ibuf[k];
+LOG_PRINTF("[%2d][%02X] ", desc_len, ibuf[k + 1]);
```
Comment (westlicht) [inline src/platform/stm32/libs/libusbhost/src/usbh_core.c:147]: weird indentation

File: src/platform/stm32/libs/libusbhost/src/usbh_core.c [hunk 3]
```diff
@@ -415,35 +416,31 @@ static void device_enumerate(usbh_device_t *dev, usbh_packet_callback_data_t cb_
...
 							(struct usb_device_descriptor *)&usbh_buffer[0];
 					dev->packet_size_max0 = ddt->bMaxPacketSize0;
-					LOG_PRINTF("Found device with vid=0x%04x pid=0x%04x\n", ddt->idVendor, ddt->idProduct);
-					LOG_PRINTF("class=0x%02x subclass=0x%02x protocol=0x%02x\n", ddt->bDeviceClass, ddt->bDeviceSubClass, ddt->bDeviceProtocol);
-					CONTINUE_WITH(USBH_ENUM_STATE_CONFIGURATION_DT_HEADER_READ_SETUP)
+					CONTINUE_WITH(USBH_ENUM_STATE_DEVICE_DT_READ_SETUP);
 				}
 				break;
-
 			case USBH_PACKET_CALLBACK_STATUS_ERRSIZ:
```
Comment (westlicht) [inline src/platform/stm32/libs/libusbhost/src/usbh_core.c:444]: Now that we only send 8 bytes (assuming that's the required minimum by the standard?!?) can we ever hit this? I guess it's still save to continue just wondering.

File: src/platform/stm32/libs/libusbhost/src/usbh_lld_stm32f4.c [hunk 4]
```diff
@@ -253,6 +253,10 @@ static void write(void *drvdata, const usbh_packet_t *packet)
 	uint32_t dpid;
 	if (packet->endpoint_type == USBH_ENDPOINT_TYPE_CONTROL) {
 		if (packet->control_type == USBH_CONTROL_TYPE_DATA) {
+			// For Status OUT stage, Length==0, Status Out PID = 1 */
+			if( !packet->datalen ) {
```
Comment (westlicht) [inline src/platform/stm32/libs/libusbhost/src/usbh_lld_stm32f4.c:257]: `if (!packet->datalen) {` according to the coding conventions I guess.

**Comment by westlicht** [comment]
@av500 sorry for the late review. I've been busy with work. Thanks a lot for these fixes. I'm not an expert in the USB protocol but I previously assumed that this library had some issues. I had some problems when trying to get the new Launchpad series to work but a firmware update from Novation seems to have resolved those issues, but it probably would have worked before with these changes in place. There are some minor comments. If you can fix those great, otherwise I'll do those changes myself.

**Comment by av500** [comment]
thx for the review, I'll check your comments

File: src/platform/stm32/libs/libusbhost/src/usbh_core.c (hunk 2 above)
Comment (av500) [reply src/platform/stm32/libs/libusbhost/src/usbh_core.c:147]: heh, in my own code all log prints are at the start :) I'll move it

File: src/platform/stm32/libs/libusbhost/src/usbh_core.c (hunk 1 above)
Comment (av500) [reply src/platform/stm32/libs/libusbhost/src/usbh_core.c:145]: right, -i is missing, adding it

File: src/platform/stm32/libs/libusbhost/src/usbh_lld_stm32f4.c (hunk 4 above)
Comment (av500) [reply src/platform/stm32/libs/libusbhost/src/usbh_lld_stm32f4.c:257]: ok

File: src/platform/stm32/libs/libusbhost/src/usbh_core.c (hunk 3 above)
Comment (av500) [reply src/platform/stm32/libs/libusbhost/src/usbh_core.c:444]: I'm not *that* deep into USB and this lib, and I don't know what weird USB devices are out there, so I guess if we get 8 bytes we should be happy and move on.

Comment (av500) [reply src/platform/stm32/libs/libusbhost/src/usbh_core.c:444]: USBH_PACKET_CALLBACK_STATUS_ERRSIZ gets set if the returned data is not *exactly* the right size, I guess that would trigger if a device returns 18 when asked for 8, so better to leave it in

**Comment by av500** [comment]
@westlicht I fixed the buffer length and the two coding style issues and re-pushed the branch

File: src/platform/stm32/libs/libusbhost/src/usbh_core.c (hunk 3 above)
Comment (westlicht) [reply src/platform/stm32/libs/libusbhost/src/usbh_core.c:444]: I agree.

**Comment by westlicht** [comment]
Looks great, merging this now.

**Comment by napoleonthegreat94** [comment]
Thank you all for fixing usb host. Should we expect a new firmware with the changes?!

**Comment by CT-030** [comment]
Hello do you have a solution with the launchpad mini mk3 ver.1.0.4 ? It isn't working :/

**Comment by Kuritana** [comment]
Nc

**Pull request merged**